File: backend/graph/category_catalog.py
"""Curated Kapruka category display catalog — mirrors kapruka.com gift grid (~39 items)."""

from __future__ import annotations

import re
from typing import Any

CategoryEntry = dict[str, Any]

# Gift-relevant categories from the Kapruka.com reference grid (39 entries).
FEATURED_CATALOG: list[CategoryEntry] = [
    {"mcp_name": "cakes", "label": "Cake Shop", "icon_slug": "cakes", "sort_order": 1},
    {"mcp_name": "combopack", "label": "Combo Gift Packs", "icon_slug": "combopack", "sort_order": 2},
    {"mcp_name": "Chocolates", "label": "Chocolates", "icon_slug": "chocolates", "sort_order": 3},
    {"mcp_name": "Clothing", "label": "Clothing", "icon_slug": "clothing", "sort_order": 4},
    {"mcp_name": "Electronic", "label": "Electronics", "icon_slug": "electronics", "sort_order": 5},
    {"mcp_name": "flowers", "label": "Flower Shop", "icon_slug": "flowers", "sort_order": 6},
    {"mcp_name": "Food", "label": "Food & Restaurants", "icon_slug": "food", "sort_order": 7},
    {"mcp_name": "Fruits", "label": "Fruit & Fruit Baskets", "icon_slug": "fruits", "sort_order": 8},
    {"mcp_name": "Vegetables", "label": "Veg & Veg Baskets", "icon_slug": "vegetables", "sort_order": 9},
    {"mcp_name": "Giftcert", "label": "Gift Vouchers & Tickets", "icon_slug": "giftcert", "sort_order": 10},
    {"mcp_name": "Giftset", "label": "Combo and Gift Sets", "icon_slug": "giftset", "sort_order": 11},
    {"mcp_name": "Grocery", "label": "Grocery Items", "icon_slug": "grocery", "sort_order": 12},
    {"mcp_name": "GreetingCards", "label": "Greeting Cards & Party", "icon_slug": "greeting_cards", "sort_order": 13},
    {
        "mcp_name": "combopack",
        "label": "Hampers",
        "icon_slug": "hampers",
        "sort_order": 14,
        "search_hint": "hampers",
    },
    {"mcp_name": "Jewellery", "label": "Jewelry & Watches", "icon_slug": "jewellery", "sort_order": 15},
    {
        "mcp_name": "Personalized Gifts",
        "label": "Personalized Gifts",
        "icon_slug": "personalized_gifts",
        "sort_order": 16,
    },
    {"mcp_name": "Perfumes", "label": "Perfumes & Fragrances", "icon_slug": "perfumes", "sort_order": 17},
    {"mcp_name": "Fashion", "label": "Hand Bags & Fashion & Shoes", "icon_slug": "fashion", "sort_order": 18},
    {"mcp_name": "Cosmetics", "label": "Cosmetics", "icon_slug": "cosmetics", "sort_order": 19},
    {"mcp_name": "Schoolpride", "label": "College Pride", "icon_slug": "schoolpride", "sort_order": 20},
    {"mcp_name": "Childrens", "label": "School Supplies", "icon_slug": "school_supplies", "sort_order": 21},
    {"mcp_name": "Books", "label": "Books", "icon_slug": "books", "sort_order": 22},
    {"mcp_name": "Pharmacy", "label": "Health and Wellness", "icon_slug": "pharmacy", "sort_order": 23},
    {"mcp_name": "KidsToys", "label": "Soft Toys & Kids Toys", "icon_slug": "kids_toys", "sort_order": 24},
    {"mcp_name": "Bicycle", "label": "Sports & Bicycles", "icon_slug": "sports", "sort_order": 25},
    {"mcp_name": "BabyItems", "label": "Mother & Baby", "icon_slug": "baby", "sort_order": 26},
    {"mcp_name": "Household", "label": "Home & Lifestyle", "icon_slug": "household", "sort_order": 27},
    {"mcp_name": "pirikara", "label": "Religious Items", "icon_slug": "pirikara", "sort_order": 28},
    {"mcp_name": "Automobile", "label": "Automobile", "icon_slug": "automobile", "sort_order": 29},
    {"mcp_name": "Pet", "label": "Petcare", "icon_slug": "pet", "sort_order": 30},
    {"mcp_name": "Adult Products", "label": "Intimate Essentials", "icon_slug": "adult", "sort_order": 31},
    {"mcp_name": "uniquegifts", "label": "Made in SL", "icon_slug": "made_in_sl", "sort_order": 32},
    {
        "mcp_name": "promotions",
        "label": "Kapruka Global Shop",
        "icon_slug": "global_shop",
        "sort_order": 33,
        "search_hint": "global shop",
    },
    {
        "mcp_name": "samedaydelivery",
        "label": "Gifts to Other Countries",
        "icon_slug": "international",
        "sort_order": 34,
        "search_hint": "international gifts",
    },
    {
        "mcp_name": "Schoolpride",
        "label": "Get SL Merchandise",
        "icon_slug": "merchandise",
        "sort_order": 35,
        "search_hint": "merchandise",
    },
    {"mcp_name": "Services", "label": "Services", "icon_slug": "services", "sort_order": 36},
    {
        "mcp_name": "Giftcert",
        "label": "Reload Mobile Phones",
        "icon_slug": "mobile_reload",
        "sort_order": 37,
        "search_hint": "mobile reload",
    },
    {
        "mcp_name": "Giftcert",
        "label": "Send Money & Vouchers",
        "icon_slug": "money_vouchers",
        "sort_order": 38,
        "search_hint": "gift voucher money",
    },
    {
        "mcp_name": "Services",
        "label": "Realestate Services",
        "icon_slug": "realestate",
        "sort_order": 39,
        "search_hint": "real estate",
    },
]

ICON_EMOJI: dict[str, str] = {
    "cakes": "🎂",
    "combopack": "🧸",
    "chocolates": "🍫",
    "clothing": "👕",
    "electronics": "📻",
    "flowers": "💐",
    "food": "🍔",
    "fruits": "🍎",
    "vegetables": "🥦",
    "giftcert": "🎫",
    "giftset": "🎁",
    "grocery": "🛒",
    "greeting_cards": "✉️",
    "hampers": "🧺",
    "jewellery": "⌚",
    "personalized_gifts": "☕",
    "perfumes": "🧴",
    "fashion": "👜",
    "cosmetics": "💄",
    "schoolpride": "🎓",
    "school_supplies": "✏️",
    "books": "📚",
    "pharmacy": "🏥",
    "kids_toys": "🧸",
    "sports": "🚲",
    "baby": "👶",
    "household": "🏠",
    "pirikara": "🛐",
    "automobile": "🏍️",
    "pet": "🐾",
    "adult": "🔞",
    "made_in_sl": "🇱🇰",
    "global_shop": "🌐",
    "international": "✈️",
    "merchandise": "👕",
    "services": "🛠️",
    "mobile_reload": "📱",
    "money_vouchers": "💸",
    "realestate": "🏘️",
}

_CATALOG_BY_MCP: dict[str, CategoryEntry] = {
    entry["mcp_name"]: entry for entry in FEATURED_CATALOG
}


def _humanize_name(name: str) -> str:
    cleaned = re.sub(r"[_\-]+", " ", name.strip())
    return cleaned[:1].upper() + cleaned[1:] if cleaned else name


def icon_url(icon_slug: str) -> str:
    return f"/categories/{icon_slug}.webp"


def catalog_entry_for(mcp_name: str) -> CategoryEntry | None:
    return _CATALOG_BY_MCP.get(mcp_name)


def catalog_label(mcp_name: str, *, fallback: str | None = None) -> str:
    entry = catalog_entry_for(mcp_name)
    if entry:
        return str(entry["label"])
    return fallback or _humanize_name(mcp_name)


def _enrich_node(node: dict[str, Any], *, catalog: CategoryEntry | None = None) -> dict[str, Any]:
    name = str(node.get("name") or "")
    entry = catalog or catalog_entry_for(name)
    label = str(entry["label"]) if entry else _humanize_name(name)
    icon_slug = str(entry["icon_slug"]) if entry else re.sub(r"\s+", "_", name.lower())
    emoji = ICON_EMOJI.get(icon_slug, "🎁")

    enriched: dict[str, Any] = {
        **node,
        "name": name,
        "label": label,
        "icon_slug": icon_slug,
        "iconUrl": icon_url(icon_slug),
        "emoji": emoji,
        "featured": entry is not None,
    }
    if entry and entry.get("search_hint"):
        enriched["search_hint"] = entry["search_hint"]

    children = node.get("children") or []
    if children:
        enriched["children"] = [
            {
                **child,
                "name": child.get("name"),
                "label": _humanize_name(str(child.get("name") or "")),
            }
            for child in children
        ]
    return enriched


def _merge_featured_with_mcp(
    mcp_categories: list[dict[str, Any]],
) -> list[dict[str, Any]]:
    by_name = {str(item.get("name") or ""): item for item in mcp_categories}
    merged: list[dict[str, Any]] = []
    for entry in sorted(FEATURED_CATALOG, key=lambda item: int(item["sort_order"])):
        mcp_node = by_name.get(entry["mcp_name"], {"name": entry["mcp_name"]})
        merged.append(_enrich_node(mcp_node, catalog=entry))
    return merged


def enrich_categories_tree(
    mcp_categories: list[dict[str, Any]],
    *,
    featured_only: bool = False,
) -> list[dict[str, Any]]:
    """Merge MCP category tree with display labels, emoji, and icon URLs."""
    if featured_only:
        return _merge_featured_with_mcp(mcp_categories)

    enriched = [_enrich_node(node) for node in mcp_categories]
    enriched.sort(
        key=lambda node: (
            not node.get("featured"),
            _CATALOG_BY_MCP.get(str(node.get("name") or ""), {}).get("sort_order", 999),
            str(node.get("label") or node.get("name") or ""),
        )
    )
    return enriched

File: frontend/src/data/categoryCatalog.ts
import type { KaprukaCategory } from "../types";

export interface CategoryMeta {
  mcpName: string;
  label: string;
  emoji: string;
  featured: boolean;
  sortOrder: number;
}

/** Curated gift categories — mirrors backend FEATURED_CATALOG for client-side enrichment. */
export const CATEGORY_CATALOG: CategoryMeta[] = [
  { mcpName: "cakes", label: "Cake Shop", emoji: "🎂", featured: true, sortOrder: 1 },
  { mcpName: "combopack", label: "Combo Gift Packs", emoji: "🎁", featured: true, sortOrder: 2 },
  { mcpName: "Chocolates", label: "Chocolates", emoji: "🍫", featured: true, sortOrder: 3 },
  { mcpName: "Clothing", label: "Clothing", emoji: "👕", featured: true, sortOrder: 4 },
  { mcpName: "Electronic", label: "Electronics", emoji: "📱", featured: true, sortOrder: 5 },
  { mcpName: "flowers", label: "Flower Shop", emoji: "💐", featured: true, sortOrder: 6 },
  { mcpName: "Food", label: "Food & Restaurants", emoji: "🍽️", featured: true, sortOrder: 7 },
  { mcpName: "Fruits", label: "Fruit Baskets", emoji: "🍎", featured: true, sortOrder: 8 },
  { mcpName: "Vegetables", label: "Veg Baskets", emoji: "🥬", featured: true, sortOrder: 9 },
  { mcpName: "Giftcert", label: "Gift Vouchers", emoji: "🎟️", featured: true, sortOrder: 10 },
  { mcpName: "Giftset", label: "Gift Sets", emoji: "🧺", featured: true, sortOrder: 11 },
  { mcpName: "Grocery", label: "Grocery", emoji: "🛒", featured: true, sortOrder: 12 },
  { mcpName: "GreetingCards", label: "Greeting Cards", emoji: "💌", featured: true, sortOrder: 13 },
  { mcpName: "Jewellery", label: "Jewelry & Watches", emoji: "💎", featured: true, sortOrder: 14 },
  { mcpName: "Personalized Gifts", label: "Personalized Gifts", emoji: "✨", featured: true, sortOrder: 15 },
  { mcpName: "Perfumes", label: "Perfumes", emoji: "🌸", featured: true, sortOrder: 16 },
  { mcpName: "Fashion", label: "Fashion & Shoes", emoji: "👗", featured: true, sortOrder: 17 },
  { mcpName: "Cosmetics", label: "Cosmetics", emoji: "💄", featured: true, sortOrder: 18 },
  { mcpName: "Books", label: "Books", emoji: "📚", featured: true, sortOrder: 19 },
  { mcpName: "Toys", label: "Toys", emoji: "🧸", featured: true, sortOrder: 20 },
  { mcpName: "baby", label: "Baby Gifts", emoji: "👶", featured: true, sortOrder: 21 },
  { mcpName: "birthday", label: "Birthday", emoji: "🎉", featured: true, sortOrder: 22 },
  { mcpName: "anniversary", label: "Anniversary", emoji: "💍", featured: true, sortOrder: 23 },
  { mcpName: "valentine", label: "Valentine", emoji: "❤️", featured: true, sortOrder: 24 },
  { mcpName: "mothersday", label: "Mother's Day", emoji: "🌷", featured: true, sortOrder: 25 },
  { mcpName: "fathersday", label: "Father's Day", emoji: "👔", featured: true, sortOrder: 26 },
  { mcpName: "Christmas", label: "Christmas", emoji: "🎄", featured: true, sortOrder: 27 },
  { mcpName: "newyear", label: "New Year", emoji: "🎆", featured: true, sortOrder: 28 },
  { mcpName: "Tea", label: "Tea & Beverages", emoji: "🍵", featured: true, sortOrder: 29 },
  { mcpName: "Health", label: "Health & Wellness", emoji: "💊", featured: true, sortOrder: 30 },
  { mcpName: "Home", label: "Home & Living", emoji: "🏠", featured: true, sortOrder: 31 },
  { mcpName: "Sports", label: "Sports", emoji: "⚽", featured: true, sortOrder: 32 },
  { mcpName: "Stationery", label: "Stationery", emoji: "✏️", featured: true, sortOrder: 33 },
  { mcpName: "plants", label: "Plants", emoji: "🪴", featured: true, sortOrder: 34 },
  { mcpName: "Pet", label: "Pet Gifts", emoji: "🐾", featured: true, sortOrder: 35 },
  { mcpName: "wesak", label: "Wesak", emoji: "🪷", featured: true, sortOrder: 36 },
  { mcpName: "deepawali", label: "Deepawali", emoji: "🪔", featured: true, sortOrder: 37 },
  { mcpName: "Musical Instruments", label: "Music", emoji: "🎵", featured: true, sortOrder: 38 },
  { mcpName: "Art", label: "Art & Crafts", emoji: "🎨", featured: true, sortOrder: 39 },
];

export const ICON_EMOJI: Record<string, string> = Object.fromEntries(
  CATEGORY_CATALOG.map((entry) => [entry.mcpName, entry.emoji]),
);

export const CATEGORY_LABELS: Record<string, string> = Object.fromEntries(
  CATEGORY_CATALOG.map((entry) => [entry.mcpName, entry.label]),
);

const catalogByName = new Map(CATEGORY_CATALOG.map((entry) => [entry.mcpName, entry]));

function titleCase(name: string): string {
  return name
    .replace(/_/g, " ")
    .replace(/\b\w/g, (char) => char.toUpperCase());
}

export function getCategoryLabel(name: string, fallback?: string): string {
  return CATEGORY_LABELS[name] ?? fallback ?? titleCase(name);
}

export function getCategoryEmoji(name: string): string {
  return ICON_EMOJI[name] ?? "🎁";
}

export function enrichCategory(category: KaprukaCategory): KaprukaCategory {
  const meta = catalogByName.get(category.name);
  return {
    ...category,
    label: category.label ?? meta?.label ?? getCategoryLabel(category.name),
    emoji: category.emoji ?? meta?.emoji ?? getCategoryEmoji(category.name),
    children: category.children?.map(enrichCategory),
  };
}

export function enrichCategories(categories: KaprukaCategory[]): KaprukaCategory[] {
  return categories.map(enrichCategory);
}

export function featuredCategories(categories: KaprukaCategory[]): KaprukaCategory[] {
  const featuredNames = new Set(
    CATEGORY_CATALOG.filter((entry) => entry.featured).map((entry) => entry.mcpName),
  );
  const enriched = enrichCategories(categories);
  const filtered = enriched.filter((cat) => featuredNames.has(cat.name));
  return filtered.sort((a, b) => {
    const orderA = catalogByName.get(a.name)?.sortOrder ?? 999;
    const orderB = catalogByName.get(b.name)?.sortOrder ?? 999;
    return orderA - orderB;
  });
}
